whisper/src/net.rs

```diff
@@ -0,0 +1,638 @@
+	/// The same message will not be passed twice.
+	/// Heavy handling should be done asynchronously.
+	/// If there is a significant overhead in this thread, then an attacker
+	/// can determine which kinds of messages we are listening for.
+	fn handle_messages(&self, message: &[Message]);
+}
+
+// errors in importing a whisper message.
+#[derive(Debug)]
+enum Error {
+	Decoder(DecoderError),
+	Network(NetworkError),
+	Message(MessageError),
+	UnknownPacket(u8),
+	UnknownPeer(PeerId),
+	ProtocolVersionMismatch(usize),
+	SameNodeKey,
+	UnexpectedMessage,
+}
+
+impl From<DecoderError> for Error {
+	fn from(err: DecoderError) -> Self {
+		Error::Decoder(err)
+	}
+}
+
+impl From<NetworkError> for Error {
+	fn from(err: NetworkError) -> Self {
+		Error::Network(err)
+	}
+}
+
+impl From<MessageError> for Error {
+	fn from(err: MessageError) -> Self {
+		Error::Message(err)
+	}
+}
+
+impl fmt::Display for Error {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		match *self {
+			Error::Decoder(ref err) => write!(f, "Failed to decode packet: {}", err),
+			Error::Network(ref err) => write!(f, "Network error: {}", err),
+			Error::Message(ref err) => write!(f, "Error decoding message: {}", err),
+			Error::UnknownPacket(ref id) => write!(f, "Unknown packet kind: {}", id),
+			Error::UnknownPeer(ref id) => write!(f, "Message received from unknown peer: {}", id),
+			Error::ProtocolVersionMismatch(ref proto) =>
+				write!(f, "Unknown protocol version: {}", proto),
+			Error::UnexpectedMessage => write!(f, "Unexpected message."),
+			Error::SameNodeKey => write!(f, "Peer and us have same node key."),
+		}
+	}
+}
+
+// sorts by work proved, descending.
+#[derive(PartialEq, Eq)]
+struct SortedEntry {
+	slab_id: usize,
+	work_proved: OrderedFloat<f64>,
+	expiry: SystemTime,
+}
+
+impl Ord for SortedEntry {
+	fn cmp(&self, other: &SortedEntry) -> Ordering {
+		self.work_proved.cmp(&other.work_proved)
+	}
+}
+
+impl PartialOrd for SortedEntry {
+	fn partial_cmp(&self, other: &SortedEntry) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+// stores messages by two metrics: expiry and PoW rating
+// when full, will accept messages above the minimum stored.
+struct Messages {
+	slab: ::slab::Slab<Message>,
+	sorted: Vec<SortedEntry>,
+	known: HashSet<H256>,
+	removed_hashes: Vec<H256>,
+	cumulative_size: usize,
+	ideal_size: usize,
+}
+
+impl Messages {
+	fn new(ideal_size: usize) -> Self {
+		Messages {
+			slab: ::slab::Slab::with_capacity(0),
+			sorted: Vec::new(),
+			known: HashSet::new(),
+			removed_hashes: Vec::new(),
+			cumulative_size: 0,
+			ideal_size: ideal_size,
+		}
+	}
+
+	// reserve space for additional elements.
+	fn reserve(&mut self, additional: usize) {
+		self.slab.reserve_exact(additional);
+		self.sorted.reserve(additional);
+		self.known.reserve(additional);
+	}
+
+	// whether a message is known or within the bounds of PoW.
+	fn may_accept(&self, message: &Message) -> bool {
+		!self.known.contains(message.hash()) && {
```

Comment:
yeah, should be is _not_ known (I had it as `is_known` before but forgot to update when I generalized it)